src/service/url-expander/expander.js

```diff
@@ -0,0 +1,264 @@
+        // Do not perform substitution and just return back the original
+        // match, so that the string doesn't change.
+        return results;
+      }
+      if (this.fillStorage_(match, keywordMarker)) {
+        // if this match does not overlap with a longer match
+        results.push(match);
+      }
+      return results;
+    }, []).sort((a, b) => a.start - b.start);
+  }
+
+  /**
+   * Given a keyword marking data structure and a match object, it will keep track of
+   * indexes that are already being tracked.
+   * @param {!Object<string, *>} match contains match data
+   * @return {boolean} indicates whether a collision was found
+   */
+  fillStorage_(match, keywordMarker) {
+    for (let i = match.start; i <= match.stop; i++) {
+      if (keywordMarker[i]) {
+        // toggle off what we thought was to be filled
+        for (let j = match.start; j < i; j++) {
+          keywordMarker[j] = false;
+        }
+        return false;
+      }
+      keywordMarker[i] = true;
+    }
+    return true;
+  }
+
+  /**
+   * @param {string} url
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {!Object<string, *>=} opt_bindings additional one-off bindings
+   * @return {!Promise<string>}
+   */
+  parseUrlRecursively_(url, matches, opt_bindings) {
+    const stack = [];
+    let urlIndex = 0;
+    let matchIndex = 0;
+    let match = matches[matchIndex];
+    let numOfPendingCalls = 0;
+
+    const evaluateNextLevel = () => {
+      let builder = '';
+      const results = [];
+
+      while (urlIndex < url.length && matchIndex <= matches.length) {
+
+        if (match && urlIndex === match.start) {
+          let binding;
+          // find out where this keyword is coming from
+          if (opt_bindings && opt_bindings.hasOwnProperty(match.name)) {
+            // the optional bindings
+            binding = opt_bindings[match.name];
+          } else {
+            // or the global source
+            binding = this.variableSource_.get(match.name);
+            binding = binding.async || binding.sync;
+          }
+
+          urlIndex = match.stop + 1;
+          match = matches[++matchIndex];
+
+          if (url[urlIndex] === '(') {
+            // if we hit a left parenthesis we still need to get args
+            urlIndex++;
+            numOfPendingCalls++;
+            stack.push(binding);
+            if (builder.length) {
+              results.push(builder);
+            }
+            results.push(evaluateNextLevel());
+          } else {
+            if (builder.length) {
+              results.push(builder);
+            }
+            results.push(this.evaluateBinding_(binding));
+          }
+
+          builder = '';
+        }
+
+        else if (numOfPendingCalls && url[urlIndex] === ',') {
+          if (builder.length) {
+            results.push(builder.trim());
+          }
+          // support legacy two comma format
```

Comment:
shouldn't be called legacy, i guess that's still the way to do it now, right?

Reply:
yeah we have to keep this to not break things. Should I just call it existing?

Reply:
just remove the wording `legacy`, cause it's still by design